### PR TITLE
[Relax] Support callback as argument

### PR DESCRIPTION
**Pull request by Lunderberg**
Prior to this commit, calls from Relax to external PackedFuncs could only be done through the TVM global registry.  While Relax functions accepting a callback could be written as `callback_arg: R.Callable(arg_struct_info, ret_struct_info)`, attempting to compile these functions would raise an error during the `CodeGenVM` step of `relax.build`.  In addition, the global registry is only queried when initializing the `relax.VirtualMachine`, and so later changes requires restarting the VM.

This commit updates both the `CodeGenVM` lowering pass and the relax VM to support callbacks.  The motivating use case is with the `LazyTransformParams` pass, to improve flexibility by avoiding use of the global registry.

### CODE REVIEW

**Comment by Lunderberg** [comment]
Merged from main to PR branch, to resolve CI breakage that was fixed in https://github.com/apache/tvm/pull/16546.

**Review by slyubomirsky — APPROVED** [review]
The change seems straightforward and it's good that there are some test cases too. I'm glad that this case will work now.

File: include/tvm/runtime/relax_vm/bytecode.h [hunk 1]
```diff
@@ -183,10 +184,15 @@ struct Instruction {
...
   Opcode op;
   union {
-    struct /* Call */ {
+    struct /* Call, CallFromRegister */ {
       /*! \brief The destination register. */
       RegName dst;
-      /*! \brief The index into the packed function table. */
+      /*! \brief The index of the function.
+       *
+       * For `OpCode::Call`, this is an index into the table of static
```
Comment (tqchen) [inline include/tvm/runtime/relax_vm/bytecode.h:192]: we should not use CallFromRegister, instead we should lower to https://github.com/apache/tvm/blob/main/src/runtime/relax_vm/builtin.cc#L373

**Comment by tqchen** [comment]
Sorry for being late on this one. In this particular case, it is better to use  `vm.builtin.invoke_closure` to lower those possible variables that contains the Callable. There are two reasons for this

- One goal of relax VM is to ensure minimal set of instructions, so far we have been keeping the miminum variant via Call and all the rest call variant using builtins like `vm.builtin.invoke_closure`
- We need to handle the case where callback is either a PackedFunc or Closure,  `invoke_closure` handles the case. This enables reuse for compiled cases

@Lunderberg  would be great to followup on this soon given this would change some binary format of the VM https://github.com/apache/tvm/blob/main/src/runtime/relax_vm/builtin.cc#L373

**Comment by tqchen** [comment]
cc @yongwww  as well

**Comment by Lunderberg** [comment]
@tqchen Thank you, and I'll make a follow-up PR.  I had thought that `vm.builtin.invoke_closure` was only for use with internally-generated closure objects.

**Comment by Lunderberg** [comment]
I've submitted https://github.com/apache/tvm/pull/16573, which updates the handling of callback functions, reverts the VM changes, and passes the unit tests that were added in this PR.

**Comment by slyubomirsky** [comment]
Good to know re: invoke_closure

**Pull request merged**